src/utils/entityStore.ts

```diff
@@ -3,95 +3,90 @@
  */
 export interface IEntityStore<T extends IPositionedEntity> {
   /**
-   * Adds an entity to the store
-   * @param {T} [entity] Entity to add to the store. Entity must be positionable
+   * Adds all of the supplied entities to the store
+   *
+   * @param {T[]} [entities] Entity array to add to the store. Entities must be
+   * positionable
    */
-  add(entity: T): void;
+  addAll(entities: T[]): void;
+
   /**
    * Returns closest entity to a given {Point}
+   *
    * @param {Point} [point] Point around which to search for a closest entity
-   * @param {(entity: T) => boolean} [filter] optional method that is called while
-   * searching for the entity nearest a point. If the filter returns false, the point
-   * is considered invalid and is not considered. If the filter returns true, the point
-   * is considered valid and will be considered.
+   * @param {(entity: T) => boolean} [filter] optional method that is called
+   * while searching for the entity nearest a point. If the filter returns
+   * false, the point is considered invalid and is not considered. If the filter
+   * returns true, the point is considered valid and will be considered.
    * @returns {T} Will return the nearest entity or undefined if none are found
    */
   entityNearest(point: Point, filter?: (entity: T) => boolean): T;
-  /**
-   * Iterator that loops through entities and returns a transformed array
-   * @param {(value: T) => S} [callback] transformation function that is passed
-   * passed an entity {T} and returns an object {S}.
-   * @returns {S[]} The aggregate result of each call to the transformation function
-   */
-  map<S>(callback: (value: T) => S): S[];
 
   /**
-   * Iterator that loops through entities calls the callback on each iteration.
-   * @param {(value: T) => void} [callback] transformation function that is passed
-   * passed an entity {T}.
+   * Returns the current internal array of entities.
+   *
+   * @returns {T[]} the current internal array of entities.
    */
-  forEach(callback: (value: T) => void): void;
-}
-
-export interface IPositionedEntity {
-  position: Point;
+  entities(): T[];
 }
 
 /**
- * Array-backed implementation of {EntityStore}
+ * Implementation of {IEntityStore} that uses an array for easy iteration as
+ * well as a quad tree for fast nearest-point queries.
+ *
+ * Note that if the position of your entities changes, you MUST rebuild the
+ * entity store for the `entityNearest` method to work since the quadtree does
+ * not know that its nodes have moved.
  */
-export class EntityArray<T extends IPositionedEntity> implements IEntityStore<T> {
+export class EntityStore<T extends IPositionedEntity> implements IEntityStore<T> {
   private _entities: T[];
+  private _tree: d3.Quadtree<T>;
 
   constructor() {
     this._entities = [];
+    this._tree = d3.quadtree<T>()
+      // int casting for faster computation. losing sub-pixel precision here is
+      // of no concern.
+      .x((d) => d.position.x | 0)
```

Comment:
nit: prefer Math.floor() to cast to int. Also do we know that casting to an int actually improves perf significantly?